clients/python-wrapper/lakefs/config.py

```diff
@@ -60,35 +107,162 @@ def __init__(self, verify_ssl: Optional[bool] = None, proxy: Optional[str] = Non
-        self.password = secret_env if secret_env is not None else self.credentials.secret_access_key
-        self.access_token = os.getenv(_LAKECTL_CREDENTIALS_SESSION_TOKEN)
+    def _has_valid_authentication(self) -> bool:
+        """Check if we have valid authentication credentials"""
+        if self.access_token is not None:
+            return True
+        if (self.username is not None and len(self.username) > 0 and
+                self.password is not None and len(self.password) > 0):
+            return True
+        if self._iam_provider is not None:
+            return True
+        return False
 
-        if self.access_token is not None or len(self.username) > 0 and len(self.password) > 0:
-            found = True
+    def get_auth_type(self) -> Optional[ClientConfig.AuthType]:
+        """
+        Returns the type of authentication used: either SessionToken, Credentials, or IAMProvider
+        :return: ClientConfig.AuthType
+        """
+        if self._iam_provider is not None:
+            return ClientConfig.AuthType.IAM
+        if self.access_token is not None:
+            return ClientConfig.AuthType.SESSION_TOKEN
+        if self.credentials is not None:
+            return ClientConfig.AuthType.CREDENTIALS
+        return None
 
-        if not found:
-            raise NoAuthenticationFound
+    def get_iam_provider(self) -> Optional[ClientConfig.IAMProvider]:
+        """
+        Returns the IAM provider used for authentication.
+        :return: ClientConfig.IAMProvider
+        """
+        if self._iam_provider is not None:
+            return self._iam_provider
+        return None
+
+def get_iam_provider_from_env_or_file(data: Optional[Dict] = None) -> Optional[ClientConfig.IAMProvider]:
+    """
+    Get IAM provider configuration from environment variables (primary) or config file (fallback).
+
+    :param data: Optional config data from YAML file
+    :return: IAMProvider if configured, None otherwise
+    """
+    provider_type = _get_provider_type(data)
+    if not provider_type:
+        return None
+
+    if provider_type not in SUPPORTED_IAM_PROVIDERS:
+        raise UnsupportedCredentialsProviderType(provider_type)
+
+    if provider_type == "aws_iam":
+        aws_config = _get_aws_iam_config(data)
+        return ClientConfig.IAMProvider(
+            type=ClientConfig.ProviderType.AWS_IAM,
+            aws_iam=aws_config
+        )
+
+    return None
+
+def _get_provider_type(data: Optional[Dict] = None) -> Optional[str]:
+    """Extract provider type from environment or config data."""
+    provider_type = os.getenv(_LAKECTL_CREDENTIALS_PROVIDER_TYPE)
+    if provider_type:
+        return provider_type
+    if data and data.get("credentials", {}).get("provider", {}).get("type"):
+        return data["credentials"]["provider"]["type"]
+
+    return None
+
+def _get_aws_iam_config(data: Optional[Dict] = None) -> ClientConfig.AWSIAMProviderConfig:
```

Comment:
This flow is really hard to follow. I'm sure you can organize it better so it will be more streamlined

Reply:
done